### PR TITLE
[MDEP-425] List plugin repositories via maven-artifact-transfer.

### PR DESCRIPTION
**Pull request by pmoerenhout**
Updated the MDEP-425 with the use of the 0.13.0 version of maven-artifact-transfer.

### CODE REVIEW

File: src/it/projects/list-plugin-repositories-verbose/pom.xml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  <name>Test</name>
+  <description>
+    Test dependency:list-plugin-repositories
+  </description>
+
+  <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  </properties>
+
```
Comment (pzygielo) [inline src/it/projects/list-plugin-repositories-verbose/pom.xml:38]: Could `pluginRepository` (I propose id of `fake-remote-repository` which is excused already in `settings.xml`) be defined in this IT and verified in test output?

File: src/main/java/org/apache/maven/plugins/dependency/AbstractDependencyMojo.java [hunk 2]
```diff
@@ -345,11 +351,25 @@ private void silenceUnarchiver( UnArchiver unArchiver )
      *         repositories, used to resolve artifacts.
      */
     public ProjectBuildingRequest newResolveArtifactProjectBuildingRequest()
+    {
+        return newProjectBuildingRequest( remoteRepositories );
+    }
+
+    /**
+     * @return Returns a new ProjectBuildingRequest populated from the current session and the current project remote
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/AbstractDependencyMojo.java:359]: nit: Returns --> returns
repositories, used --> Repositories. Used

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+                printClassesFromArtifactResult( result );
+            }
+        }
+        catch ( ArtifactResolverException | DependencyResolverException | IOException e )
+        {
+            throw new MojoExecutionException( "Couldn't download artifact: " + e.getMessage(), e );
+        }
+    }
+
+    private void printClassesFromArtifactResult( ArtifactResult result )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:186]: delete "FromArtifactResult" as its redundant with the argument

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+            while ( entries.hasMoreElements() )
+            {
+                JarEntry entry = (JarEntry) entries.nextElement();
+                String entryName = entry.getName();
+
+                // filter out files that do not end in .class
+                if ( !entryName.endsWith( ".class" ) )
+                {
+                    continue;
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:202]: consider reversing the if and rewriting without continue

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+                        + "groupId:artifactId:version[:packaging[:classifier]] " + artifact );
+            }
+            coordinate.setGroupId( tokens[0] );
+            coordinate.setArtifactId( tokens[1] );
+            coordinate.setVersion( tokens[2] );
+            if ( tokens.length >= 4 )
+            {
+                coordinate.setType( tokens[3] );
+            }
+            if ( tokens.length == 5 )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:235]: >= for safety

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+        repositorySystem.injectMirror( repoList, settings.getMirrors() );
+        repositorySystem.injectProxy( repoList, settings.getProxies() );
+        repositorySystem.injectAuthentication( repoList, settings.getServers() );
+
+        buildingRequest.setRemoteRepositories( repoList );
+
+        return buildingRequest;
+    }
+
+    protected ArtifactCoordinate toArtifactCoordinate( DependableCoordinate dependableCoordinate )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:275]: why protected?

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+            sb.append( " location: " ).append( location ).append( System.lineSeparator() );
+        }
+        this.getLog().info( sb.toString() );
+    }
+
+    /**
+     * Parses the given String into GAV artifact coordinate information, adding the given type.
+     *
+     * @param artifactString should respect the format <code>groupId:artifactId[:version]</code>
+     * @param type The extension for the artifact, must not be <code>null</code>.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:152]: nit: The --> the and no period, per Oracle JavaDoc guidelines

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+     * @return the <code>Artifact</code> object for the <code>artifactString</code> parameter.
+     * @throws MojoExecutionException if the <code>artifactString</code> doesn't respect the format.
+     */
+    private ArtifactCoordinate getArtifactCoordinate( String artifactString, String type )
+        throws MojoExecutionException
+    {
+        if ( org.codehaus.plexus.util.StringUtils.isEmpty( artifactString ) )
+        {
+            throw new IllegalArgumentException( "artifact parameter could not be empty" );
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:161]: could not be --> is

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+     *
+     * @param artifactString should respect the format <code>groupId:artifactId[:version]</code>
+     * @param type The extension for the artifact, must not be <code>null</code>.
+     * @return the <code>Artifact</code> object for the <code>artifactString</code> parameter.
+     * @throws MojoExecutionException if the <code>artifactString</code> doesn't respect the format.
+     */
+    private ArtifactCoordinate getArtifactCoordinate( String artifactString, String type )
+        throws MojoExecutionException
+    {
+        if ( org.codehaus.plexus.util.StringUtils.isEmpty( artifactString ) )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:159]: Please use apache commons stringutils instead

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        return coordinate;
+    }
+
+    /**
+     * Retrieves the Maven Project associated with the given artifact String, in the form of
+     * <code>groupId:artifactId[:version]</code>. This resolves the POM artifact at those coordinates and then builds
+     * the Maven project from it.
+     *
+     * @param artifactString Coordinates of the Maven project to get.
+     * @return New Maven project.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:204]: no period, no cap

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+     * @throws MojoExecutionException If there was an error while getting the Maven project.
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:205]: no period, no cap

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+        {
+            ProjectBuildingRequest pbr = new DefaultProjectBuildingRequest( session.getProjectBuildingRequest() );
+            pbr.setRemoteRepositories( pluginRepositories );
+            pbr.setProject( null );
+            pbr.setValidationLevel( ModelBuildingRequest.VALIDATION_LEVEL_MINIMAL );
+            pbr.setResolveDependencies( false );
+            Artifact artifact = artifactResolver.resolveArtifact( pbr, coordinate ).getArtifact();
+            return projectBuilder.build( artifact.getFile(), pbr ).getProject();
+        }
+        catch ( Exception e )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java:221]: avoid catching Exception. Can this be more specific?

File: src/site/apt/usage.apt.vm [hunk 13]
```diff
@@ -688,6 +688,24 @@ mvn dependency:build-classpath -Dmdep.outputFile=cp.txt
   poms and declared in transitive dependency poms.
 
 
+* <<<dependency:list-repositories>>>
+
+  This goal is used to list all the plugin repositories that this build depends upon. It will show plugin repositories defined in your settings,
```
Comment (elharo) [inline src/site/apt/usage.apt.vm:693]: is used to list --> lists
will show --> shows

File: src/site/apt/usage.apt.vm [hunk 14]
```diff
@@ -688,6 +688,24 @@ mvn dependency:build-classpath -Dmdep.outputFile=cp.txt
...
+  poms and declared in transitive dependency poms.
+
+  This goal can be executed from the command line:
+
++-----+
+mvn dependency:list-plugin-repositories
++-----+
+
+  Optionally, in verbose or debug mode it will display the location of the listed plugin repository:
```
Comment (elharo) [inline src/site/apt/usage.apt.vm:702]: will display --> displays
(Tech writing lives in the eternal present.)

File: src/test/java/org/apache/maven/plugins/dependency/AbstractDependencyMojoTest.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    static class ConcreteDependencyMojo extends AbstractDependencyMojo
+    {
+        static ConcreteDependencyMojo createConcreteDependencyMojoWithArtifactRepositories(
+                MavenSession mavenSession, List<ArtifactRepository> artifactRepos )
+                throws NoSuchFieldException, IllegalAccessException
+        {
+            ConcreteDependencyMojo cdm = new ConcreteDependencyMojo();
+            cdm.session = mavenSession;
+
+            Field par = AbstractDependencyMojo.class.getDeclaredField( "remoteRepositories" );
```
Comment (elharo) [inline src/test/java/org/apache/maven/plugins/dependency/AbstractDependencyMojoTest.java:55]: what is "par"? perhaps a more descriptive name?

File: src/main/java/org/apache/maven/plugins/dependency/AbstractDependencyMojo.java (hunk 2 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/plugins/dependency/AbstractDependencyMojo.java:359]: Implemented correct JavaDoc

File: src/site/apt/usage.apt.vm (hunk 13 above)
Comment (pmoerenhout) [reply src/site/apt/usage.apt.vm]: Implemented correct English...

File: src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java (hunk 12 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/plugins/dependency/resolvers/ListPluginRepositoriesMojo.java]: Catch thrown exceptions.

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java (hunk 6 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:275]: Added via rebase of master. Should it be changed in this branch?

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java (hunk 5 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:235]: Added via rebase of master. Should it be changed in this branch?

File: src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java (hunk 4 above)
Comment (pmoerenhout) [reply src/main/java/org/apache/maven/plugins/dependency/ListClassesMojo.java:202]: Added via rebase of master. Should it be changed in this branch?

File: src/site/apt/usage.apt.vm (hunk 14 above)
Comment (pmoerenhout) [reply src/site/apt/usage.apt.vm]: Changed.

**Comment by slawekjaranowski** [comment]
@pmoerenhout and other thanks for efforts ... we should not use maven-artifact-transfer at all 
It should be reimplemented with using direct resolver API like I did in #301

So I close it now.

**Comment by jira-importer** [comment]
Resolve #868

**Pull request closed without merging**